lib/core/utils/get-selector.js

```diff
@@ -1,104 +1,218 @@
+    	document.querySelectorAll(id).length === 1
+    ) {
+      return id;
+    }
+  },
+  // Get custom element name
+  getCustomElm (elm, { isCustomElm, nodeName }) {
+    if (isCustomElm) {
+      return nodeName;
+    }
+  },
+
+  // Get ARIA role
+  getElmRoleProp (elm) {
+    if (elm.hasAttribute('role')) {
+      return '[role="' + escapeSelector(elm.getAttribute('role')) +'"]';
+    }
+  },
+  // Get uncommon node names
+  getUncommonElm (elm, { isCommonElm, isCustomElm, nodeName })  {
+    if (!isCommonElm && !isCustomElm) {
+    	nodeName = escapeSelector(nodeName);
+      // Add [type] if nodeName is an input element
+      if (nodeName === 'input' && elm.hasAttribute('type')) {
+        nodeName += '[type="' + elm.type + '"]';
+      }
+      return nodeName;
+    }
+  },
+  // Has a name property, but no ID (Think input fields)
+  getElmNameProp (elm) {
+    if (!elm.id && elm.name) {
+      return '[name="' + escapeSelector(elm.name) + '"]';
+    }
+  },
+  // Get any distinct classes (as long as there aren't more then 3 of them)
+  getDistinctClass (elm, { distinctClassList }) {
+    if (distinctClassList.length > 0 && distinctClassList.length < 3) {
+      return '.' + distinctClassList.map(escapeSelector).join('.');
+    }
+  },
+  // Get a selector that uses src/href props
+  getFileRefProp (elm) {
+    let attr;
+    if (elm.hasAttribute('href')) {
+      attr = 'href';
+    } else if (elm.hasAttribute('src')) {
+      attr = 'src';
+    } else {
+      return;
+    }
+    const friendlyUriEnd = axe.utils.getFriendlyUriEnd(elm.getAttribute(attr));
+    if (friendlyUriEnd) {
+      return '[' + attr + '$="' + encodeURI(friendlyUriEnd) + '"]';
+    }
+  },
+  // Get common node names
+  getCommonName (elm, { nodeName, isCommonElm }) {
+    if (isCommonElm) {
+      return nodeName;
+    }
+  }
+};
 
-	for (index = 0; index < length; index++) {
-		sibling = siblings[index];
-		if (sibling !== node && axe.utils.matchesSelector(sibling, selector)) {
-			return true;
-		}
-	}
-	return false;
+/**
+ * Get an array of features (as CSS selectors) that describe an element
+ *
+ * By going down the list of most to least prominent element features,
+ * we attempt to find those features that a dev is most likely to
+ * recognize the element by (IDs, aria roles, custom element names, etc.)
+ */
+function getElmFeatures (elm, featureCount) {
+  const nodeName = elm.nodeName.toLowerCase();
+  const classList = Array.from(elm.classList) || [];
+  // Collect some props we need to build the selector
+  const props = {
+    nodeName,
+    classList,
+    isCustomElm: nodeName.includes('-'),
+    isCommonElm: commonNodes.includes(nodeName),
+    distinctClassList: getDistinctClassList(elm)
+  };
+
+  return [
+    // go through feature selectors in order of priority
+    createSelector.getCustomElm,
+    createSelector.getElmRoleProp,
+    createSelector.getUncommonElm,
+    createSelector.getElmNameProp,
+    createSelector.getDistinctClass,
```

Comment:
Should we demote classes even further? They often get added after events like focus.